book/src/confidence.md

```diff
@@ -0,0 +1,97 @@
+
+An entry `p` at index `i` in the `Confidence` vector implies that the validator
+observed `p` percentage of the network reaching `i` confirmations on block `N`
+as observed in some block `M`. There will be `MAX_LOCKOUT` elements in this 
+vector, representing all the possible number of confirmations from 1 and to
+`MAX_LOCKOUT`.
+
+# Computation of confidence metric for banks >= root
+
+Builiding this `Confidence` vector leverages the computations already being
+performed for building consensus. The `collect_vote_lockouts` function in 
+`consensus.rs` builds a HashMap, where each entry is of the form `(b, s)`
+where `s` is a `StakeLockout` struct representing the amount of stake and 
+lockout on a bank `b`.
+
+This computation is performed on a votable candidate bank `b` as follows.
+
+```
+   let output: HashMap<b, StakeLockout> = HashMap::new();
+   for vote_account in b.vote_accounts {
+       for v in vote_account.vote_stack {
+           for a in ancestors(v) {
+               f(*output.get_mut(a), vote_account, v);
+           }
+       }
+   }
+```
+
+where `f` is some accumulation function that modifies the `StakeLockout` entry
+for slot `a` with some data derivable from vote `v` and `vote_account`
+(stake, lockout, etc.). Note here that the `ancestors` here only includes 
+slots >= the current root. 
+
+Now we can naturally augment the above computation to also build a `Confidence`
+vector for every bank `b` by:
+1) Augmenting `StakeLockout` to include a `confidence` field that represents a
+`Confidence` vector
+2) Replacing `f` with `f'` such that the above computation also builds this
+`Confidence` vector for every bank `b`.
+
+We will proceed with the details of 2) as 1) is trivial.
+
+Before continuing, it is noteworthy that for some validator's vote account `a`,
+the number of local confirmations for that validator on slot `s` is 
+`v.num_confirmations`, where `v` is the smallest vote in the stack of votes 
+`a.votes` such that `v.slot >= s` (i.e. there is no need to look at any 
+votes > v as the number of confirmations will be lower).
+
+Now more specifically, we augment the above computation to:
+
+```
+   let output: HashMap<b, StakeLockout> = HashMap::new();
+   for vote_account in b.vote_accounts {
+       // vote stack is sorted from oldest vote to newest vote
+       for (v1, v2) in vote_account.vote_stack.windows(2) {
+           for a in ancestors(v1).difference(ancestors(v2)) {
+               f'(*output.get_mut(a), vote_account, v);
+           }
+       }
+   }
+```
+
+where `f'` is defined as:
+```
+    fn f`(
+        some_ancestor: &mut StakeLockout, 
+        vote_account: VoteAccount, 
+        v: Vote, total_stake: u64
+    ){
+        f(some_ancestor, vote_account, v);
+        *some_ancestor.confidence[v.num_confirmations] += vote_account.stake / total_stake;
+    }
+```
+
+# Computation of confidence metric for banks < root
+In the above section, the `output` HashMap computed will only include entries
```

Comment:
Some ideas @aeyakovenko and I considered while in SF:

1) Create a separate type of response if a client queries for a signature in a slot `s` < root vs the `Confidence` vector we respond with when querying for a slot >= root. In this response, a validator replies with the percentage of the network that has also set a root greater than `s`. 

This approach might lead to jarring responses for a client that is constantly querying the status of the block. They get one type of response for a while (the `Confidence` vector), and then suddenly they get another type of response, which in many ways is much less informative. 

For instance, let's say a client is waiting on at least 60% of the network to reach 7 confirmations on a block `s`. On the first response the client sees 50% of the network has reached 7 confirmations, then 55%, but before they reach 60%  the validator they were querying sets a root > `s`, and now the client sees that only 10% of the network has reached root > `s`. They will now have to adapt their decision criteria to handle this.

2.

Reply:
@carllin take a look at #6327, the probability of a fork being accepted depends on the likelihood of a streak that doubles the lockout within the lockout.